### PR TITLE
implement RFC-402

### PR DESCRIPTION
**Pull request by slangmgh**
https://github.com/nim-lang/RFCs/issues/402

### CODE REVIEW

**Comment by slangmgh** [comment]
The error msg is not matched with the previous behaviour.
So the tsigmatch.nim check failed.

**Comment by timotheecour** [comment]
can the last param be checked specially so that we allow optional params with an untyped body:
```nim
when defined case6:
  template fn(a = 1, b = 2, body) = discard

  fn(1, 2):
    foo1
    foo2

  fn(1): # bug: doesn't work
    foo1
    foo2
```

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -2346,6 +2388,14 @@ proc matchesAux(c: PContext, n, nOrig: PNode, m: var TCandidate, marker: var Int
     formal = if formalLen > 1: m.callee.n[1].sym else: nil # current routine parameter
     container: PNode = nil # constructed container
 
+  # Check if the argument count match with the parameter count,
+  # if doesn't match, fail fast
+  # See https://github.com/nim-lang/RFCs/issues/402
+
```
Comment (timotheecour) [inline compiler/sigmatch.nim:2394]: please add tests to ensure this works and at least doesnt' introduce regressions

things to consider:
* varags[int]
* varags[untyped]
* optional params
etc

**Comment by slangmgh** [comment]
> can the last param be checked specially so that we allow optional params with an untyped body:
> 
> ```nim
> when defined case6:
>   template fn(a = 1, b = 2, body) = discard
> 
>   fn(1, 2):
>     foo1
>     foo2
> 
>   fn(1): # bug: doesn't work
>     foo1
>     foo2
> ```

You can define another overloaded template or using macro with varargs[untyped].

**Comment by solo989** [comment]
Making default arguments work with untyped bodies sounds like it should be part of a completely seperate RFC.

Possibly combined with named only arguments support as a pragma or a new postfix operator so it doesn't break existing code.

**Comment by metagn** [comment]
Is there anything to do here besides test for `varargs[untyped]`?

**Pull request closed without merging**